extensions/typescript-language-features/src/typescriptServiceClient.ts

```diff
@@ -973,6 +975,46 @@ export default class TypeScriptServiceClient extends Disposable implements IType
 			case EventName.projectLoadingFinish:
 				this.loadingIndicator.finishedLoadingProject((event as Proto.ProjectLoadingFinishEvent).body.projectName);
 				break;
+
+			case EventName.createDirectoryWatcher:
+				this.createFileSystemWatcher(event.body.id, new vscode.RelativePattern(vscode.Uri.file(event.body.path), event.body.recursive ? '**' : '*'), [ /* TODO need to fill in excludes list */]);
```

Comment:
@sheetalkamat fyi this is still required to make it work: we need to know what things to exclude, otherwise VS Code will refuse to accept this request for watching.

Reply:
I am not sure how to write exclude pattern here. We know things that should definitely be included (js, mjs. cjs, ts, mts, cts json and any dts files which is .d.someExtension) There is also capability to support extra extensions and that list is given by `protocol.ConfigureRequestArguments` so those are also something we care.

Reply:
But here is the list of things we definitely ignore..  If the file path has `/node_modules/.`,  `/.git` or `/.#` we definitely ignore. 
Will adding that pattern be enough instead

Reply:
Also if the directory watcher is not recursive, do we need exclude pattern and new API usage ?

Reply:
@sheetalkamat yes, you are right, only **recursive** watchers require `exclude: string[]` to leverage the new capability, you do not have to define excludes for non-recursive watchers.

For the recursive watcher we actually get some nice extended glob capabilities because the glob pattern is directly interpreted in the native watcher and that one uses `micromatch`. With `micromatch` you can prefix patterns with `!` to negate them, so you can essentially do `excludes: ["!**/*.ts"]` to only include `*.ts` files. I am quite sure you can build a nice pattern for the files you are interested in with that approach.

However, another approach to above would be to not use `*` or `**` as pattern for the `createFileSystemWatcher(pattern)`. You can specify a glob pattern there as well to directly indicate the files you want to be notified about. Think about it like the counterpart of `exclude: string[]`, but for includes.

Note that glob patterns used in `pattern` of `createFileSystemWatcher(pattern)` and as `exclude: string[]` for non-recursive watchers are slightly less powerful. Specifically negates are not supported. See https://code.visualstudio.com/docs/editor/glob-patterns for reference.

Reply:
Here is my suggestion:
* recursive watching
  * define `excludes` leveraging the [extended capabilities](https://github.com/micromatch/micromatch#matching-features) from `micromatch` with leading `!`
  * instead of `*`, pass in a pattern for the files you care about following our [glob reference](https://code.visualstudio.com/docs/editor/glob-patterns)
* non-recursive watching
  * leave `excludes` as `undefined`
  * instead of `*`, pass in a pattern for the files you care about following our [glob reference](https://code.visualstudio.com/docs/editor/glob-patterns)

Reply:
I thought about this. but there is no safe way to exclude any extension especially because we have language plugins that can handle the additional file extensions without informing typescript about it. So the exclude pattern would have to be paths of certain ignored patterns as mentioned in https://github.com/microsoft/vscode/pull/193848#discussion_r1357164773

Reply:
Yeah I think you have to come up with the list that works best for you, but the list cannot be empty currently so that you opt-in to the new behaviour.